Frameworks/ImageIO/CGImageDestination.mm

```diff
@@ -28,75 +885,358 @@ CGImageDestinationRef CGImageDestinationCreateWithDataConsumer(CGDataConsumerRef
+
+    // Set size and resolution of the image frame
+    status = imageBitmapFrame->SetSize(imageWidth, imageHeight);
+    if (!SUCCEEDED(status)) {
+        NSTraceInfo(TAG, @"Set Frame Size failed with status=%x\n", status);
+        return;
+    }
+
+    // TIFF on iOS will only set a resolution if both DPIWidth and DPIHeight are specified, with no default values
+    // This gets written as part of writing the frame metadata, so no else part here.
+    if (imageDestination.type != typeTIFF) {
+        // Using resolution of 72 dpi as standard for now, this seems to only affect metadata and not the image itself
+        status = imageBitmapFrame->SetResolution(72.0, 72.0);
+        if (!SUCCEEDED(status)) {
+            NSTraceInfo(TAG, @"Set Frame Resolution failed with status=%x\n", status);
+            return;
+        }
+    }
+
+    // Setting up writing properties to individual image frame, bitmaps cannot have metadata
+    ComPtr<IWICMetadataQueryWriter> imageFrameMetadataWriter;
+    if (imageDestination.type != typeBMP) {
+        status = imageBitmapFrame->GetMetadataQueryWriter(&imageFrameMetadataWriter);
+        if (!SUCCEEDED(status)) {
+            NSTraceInfo(TAG, @"Get Frame Metadata Writer failed with status=%x\n", status);
+            return;
+        }
+    }
+
+    IWICMetadataQueryWriter* propertyWriter = imageFrameMetadataWriter.Get();
+
+    // Set the pixel format based on file format and write necessary metadata
+    // The bitmap frame is passed in for TIFF because WIC needs to set the IFD resolution/DPI fields through SetResolution.
+    WICPixelFormatGUID formatGUID;
+    switch (imageDestination.type) {
+        case typeJPEG:
+            formatGUID = GUID_WICPixelFormat24bppBGR;
+            writeJPEGProperties(propertyWriter, properties, imageWidth, imageHeight);
+            break;
+        case typeTIFF:
+            formatGUID = GUID_WICPixelFormat32bppRGBA;
+            writeTIFFProperties(propertyWriter, properties, imageWidth, imageHeight, imageBitmapFrame.Get());
+            break;
+        case typeGIF:
+            formatGUID = GUID_WICPixelFormat8bppIndexed;
+            writeGIFProperties(propertyWriter, properties, imageWidth, imageHeight);
+            break;
+        case typePNG:
+            formatGUID = GUID_WICPixelFormat32bppRGBA;
+            writePNGProperties(propertyWriter, properties, imageWidth, imageHeight);
+            break;
+        case typeBMP:
+            formatGUID = GUID_WICPixelFormat32bppRGBA;
+            break;
+        default:
+            NSTraceInfo(TAG, @"Unknown type encountered\n");
+            return;
+    }
+
+    status = imageBitmapFrame->SetPixelFormat(&formatGUID);
+    if (!SUCCEEDED(status)) {
+        NSTraceInfo(TAG, @"Set Pixel Format failed with status=%x\n", status);
+        return;
+    }
+
+    CGDataProviderRef provider = CGImageGetDataProvider(image);
+    NSData* imageByteData = (id)CGDataProviderCopyData(provider);
```

Comment:
As far as I can tell, everyone online says that the way to get the bytes out of a CGImageRef object is to copy it to an array using CGDataProviderCopyData. I then have to call the imageFactory->CreateBitmapFromMemory to create a bitmap that WIC can use. I might be wrong but I'm not sure if this can be avoided since there doesn't seem to be a direct way to access the bytes in a CGImageRef.